### PR TITLE
Allowing granular access for data-stream related transport actions

### PR DESCRIPTION
**Pull request by sandeshkr419**
Signed-off-by: Sandesh Kumar <sandeshkr419@gmail.com>

##  opendistro-for-elasticsearch/security pull request intake form

 
 1. __Category:__ New feature
 
 
 
 2. __Github Issue #1168__

 3. __Description of changes:__ Datastream support for OS is in progress: https://github.com/opensearch-project/OpenSearch/issues/675 
With this PR, we want to ensure that granular access for data stream related transport actions can be provided from security plugin.
This PR has a dependency on the above OS issue to be resolved and the linked PR to be merged to OS.

   The 4 transport actions that we use to access for data stream related operations in OpenSearch:
   (1).  indices:admin/data_stream/create
   (2). indices:admin/data_stream/get
   (3). indices:admin/data_stream/delete
   (4). indices:monitor/data_stream/stats

* Fixed the index resolution for (1)
* Added unit tests for all 4 transport actions.

 4. __Why these changes are required?__ 
To enable granular access for above mentioned transport actions by security plugin.

 5. __What is the old behavior before changes and new behavior after changes?__ 
Transport action (1) does not resolves correctly. (Details in github issue)

 6. __Testing done:__ 
 * Unit tests pass now.
 * Tested the transport action behavior with users by giving them multiple index-patterns in permission: 'pattern*', 'pattern', '*'. Under all these permissions, the data stream creation happens as per the index pattern specified.
 
 
 
 7. __TO-DOs, if any:__ No

 8. __Is it backport from main branch?__ No

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by sujithvm** [comment]
Please fix the failing CI

**Comment by sandeshkr419** [comment]
The core artifacts are ready in `-rc1` and were required for my changes to build and CI to pass. Therefore, changing adding the `version_qualifier` as discussed in https://github.com/opensearch-project/security/pull/1174

Also, removed the `-beta1` from OpenSearch version in [pom.xml](https://github.com/opensearch-project/security/blob/main/pom.xml#L71)

**Comment by cliu123** [comment]
> The core artifacts are ready in `-rc1` and were required for my changes to build and CI to pass. Therefore, adding the `version_qualifier` as discussed in #1174
> 
> Also, removed the `-beta1` from OpenSearch version in [pom.xml](https://github.com/opensearch-project/security/blob/main/pom.xml#L71)

We’ll have the CI changes in the other PR: https://github.com/opensearch-project/security/pull/1179. Please wait on #1179 to be merged.

**Comment by sandeshkr419** [comment]
> > The core artifacts are ready in `-rc1` and were required for my changes to build and CI to pass. Therefore, adding the `version_qualifier` as discussed in #1174
> > Also, removed the `-beta1` from OpenSearch version in [pom.xml](https://github.com/opensearch-project/security/blob/main/pom.xml#L71)
> 
> We’ll have the CI changes in the other PR: #1179. Please wait on #1179 to be merged.

Thanks Chang for the response!
I will re-base from main branch once #1179 is merged. That will remove any CI related changes in this PR.

**Comment by sandeshkr419** [comment]
#1179 is merged. 
Merged from the main branch to get the required CI related changes.
No CI related changes are part of this PR now.

**Review by sujithvm — APPROVED** [review]

File: src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java
Comment (vrozov) [inline src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java:665]: Should it be similar to `CreateIndexRequest` handling?

File: src/test/java/org/opensearch/security/DataStreamIntegrationTests.java
Comment (vrozov) [inline src/test/java/org/opensearch/security/DataStreamIntegrationTests.java:11]: Missing license header

Comment (sandeshkr419) [reply src/test/java/org/opensearch/security/DataStreamIntegrationTests.java:15]: Corrected and committed!

File: src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java
Comment (vrozov) [reply src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java:665]: Checked with @sandeshkr419: `CreateDataStreamAction.Request` unlike `CreateIndexRequest` does not support request index replace, so it should be handled similar to `BulkShardRequest`, not `CreateIndexRequest`. 

@sandeshkr419 even though the last parameter won't be used, can you change it to `false` and add comment explaining that the request is handled similar to `BulkShardRequest`.

Comment (sandeshkr419) [reply src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java]: Referring to BulkShardRequest can add to confusion here. 

To explain it better, the only difference between handling of `CreateIndexRequest` & `CreateDataStreamAction.Request` is that the latter does not supports index/data-stream replace functionality for now and therefore the analogous call to replace index (like in `CreateIndexRequest`) is not made in `CreateDataStreamAction.Request`.

I have changed the last parameter to false now.

File: src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java
Comment (vrozov) [inline src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java:665]:
```suggestion
            String[] newDataStreams = provider.provide(((CreateDataStreamAction.Request) request).indices(), request, false);
```

**Review by vrozov — APPROVED** [review]
LGTM

**Review by cliu123 — APPROVED** [review]

**Pull request merged**
